src/Npgsql/PreparedStatementManager.cs

```diff
@@ -98,161 +96,180 @@ internal PreparedStatementManager(NpgsqlConnector connector)
     internal PreparedStatement? TryGetAutoPrepared(NpgsqlBatchCommand batchCommand)
     {
         var sql = batchCommand.FinalCommandText!;
-        if (!BySql.TryGetValue(sql, out var pStatement))
+        // We could also test for PreparedState.BeingPrepared as it's handled the exact same way as PreparedState.Prepared
+        // But since it's so rare we'll just go through the slow path
+        if (!BySql.TryGetValue(sql, out var pStatement) || pStatement.State != PreparedState.Prepared)
```

Comment:
Curiously, we don't specify a comparer for `BySql` dictionary. Was there a reason for this @roji?

Reply:
What would be better than EqualityComparer.Default here?

Reply:
Isn't ordinal supposed to be faster on the account of not handling culture peculiarities?

Reply:
https://github.com/dotnet/runtime/issues/29714

Reply:
Yikes, now I have to go through all of the 9 year old codebase to remove ordinal...